### DISCUSSION TITLE
[Idea] Dropdown list separators?

### DISCUSSION THREAD

**Discussion by The-Gig** [torhus/monster-browser#51] [Ideas]
A small "problem" with MonsterBrowser is that Games and Mods are shown as the same... thinking about a simple workaround, would it be possible to allow to add "separators" through mods.ini, so it would be easier to recognize mods for a certain game in the dropdown list?
Something like adding in mods.ini a blank section (with just the brackets [] or with a space [ ] or with a minus [-] or with an underscore [_], what you prefer), which would be shown as a separator line in the dropdown list?

To get something like this
![immagine](https://user-images.githubusercontent.com/23258514/200781927-10c99735-1ae7-418a-9c16-e8c357b9baf1.png)
(a separator between Quake3 -the only one which currently has got mods out-of-the-box- and the other games)
or like this
![immagine](https://user-images.githubusercontent.com/23258514/200780970-6aa15a61-abf8-4b08-8817-8dda0120084a.png)
(a separator between each game, as one might add mods to other games)

Now you will be saying "but people can already do that, just like it you did for the example!" (so we could just add a few "default" separators to a next release and call it a day)... to which I answer saying that:
- Leaving the number of "minus characters" to the user can bring to different lenghts. I think it would be better to use a specific fixed code which the program would automatically translate into a line.
- The fact that you can actually _select_ one of those blank lines isn't the most elegant thing ever. It would be nice if somehow those separator lines would be not clickable, or clicking on them would automatically select the line _just below_ them.

Further notes:
- The number of lines shown in the dropdown list (8 at the moment) should be enlarged a bit to compensate (IDK, maybe to 12 in the case of the first example and to 16-18 in case of the second one?)
- If one makes the separator lines not selectable, then if one would type any parameters (such as "mod=whatever") under a blank section, they would just be ignored, I guess.

**Comment by torhus** [comment]
Good idea, thanks! I added a separator and made the combo box tall enough to show all the games, looks much better now :)

![New_game_list](https://user-images.githubusercontent.com/19881605/200940709-5b75fbed-f397-4ab5-8d9b-b44e14313ce1.png)

https://ci.appveyor.com/project/torhus/monster-browser/builds/45333249/artifacts

**Reply by The-Gig** [reply]
Good! I'll test it later...

By the way, a random thought: the text above the combobox says "Game"... What about "Game/Mod"?
Just an idea...

**Reply by The-Gig** [reply]
I tested it, but it looks like it's still the "crude" thing: the separator lenght depends from how many minus the user types, and it can be selected (resulting in actual results if you add some settings inside the section).
Better than nothing, but I hoped a bit more "elegant" solution was somehow possible... I guess that may imply to mess with Windows standard mechanisms? Are you sure your developing environment doesn't support "standardized" separators? (Example: <https://stackoverflow.com/questions/18154167/how-to-add-a-separator-to-a-wpf-combobox-that-is-databound>, which shows how to do in C# by tweaking borders, but unfortunately without loading the data from a file. And IIRC you are using D instead...)

**Reply by torhus** [reply]
I think "crude" is ok, it matches how it looks in mods.ini. SWT doesn't have an obvious way to add a separator, and I couldn't see any combo boxes with separators in Eclipse. So it probably requires a custom implementation of the whole combo box. CCombo doesn't have it either, but I guess that could be a starting point if I want to do this in the future.

https://stackoverflow.com/questions/5339658/how-to-add-an-unselectable-separator-to-a-combobox-in-swt

I made it so that selecting the separator doesn't actually work, at least. A section name consisting of only dashes is treated specially.

**Reply by torhus** [reply]
> By the way, a random thought: the text above the combobox says "Game"... What about "Game/Mod"? Just an idea...

Maybe, I just like that it's so short and simple now...

**Reply by torhus** [reply]
> By the way, a random thought: the text above the combobox says "Game"... What about "Game/Mod"? Just an idea...

I tried changing it to "Games and mods", and it just looks right to me 👍

**Reply by The-Gig** [reply]
Hi! I tried MB version 0.9i...
So I've seen what you meant with "selecting the separator doesn't actually work": the program automatically falls back to the previously chosen game. Very well, thank you!
The number of dashes shown still depends from how many minus signs the user typed in mods.ini, but the most important part is done!
